proto_proxy.c

```diff
@@ -19,17 +19,28 @@ static void *mcp_profile_alloc(void *ud, void *ptr, size_t osize, size_t nsize);
 /******** EXTERNAL FUNCTIONS ******/
 // functions starting with _ are breakouts for the public functions.
 
-// We want to ensure we slowly iterate through VM memory in two cases:
-// - If using the newer zero-allocation API, old backends will never close
-// after reload.
-// - If the server is completely idle and user is issuing reloads, they will
-// be confused when old backends do not close.
-// So every few seconds we push one minimal GC step.
-static void proxy_gc_poke(evutil_socket_t fd, short event, void *arg) {
-    LIBEVENT_THREAD *t = arg;
+// Attempting to push the lua GC algorithm to run outside of when requests are
+// being processed.
+void proxy_gc_poke(LIBEVENT_THREAD *t) {
     lua_State *L = t->L;
+    int vm_kb = lua_gc(L, LUA_GCCOUNT) + t->proxy_vm_extra_kb;
+    // only add GC debt with what we grew past the last reload size.
+    if (vm_kb > t->proxy_vm_last_kb * 1.5) {
+        // reset our counter once a full GC has completed
+        if (lua_gc(L, LUA_GCSTEP, vm_kb - t->proxy_vm_last_kb)) {
+            t->proxy_vm_last_kb = lua_gc(L, LUA_GCCOUNT);
+            t->proxy_vm_extra_kb = 0;
+        }
+    }
+}
+
+// every couple seconds we force-run one GC step.
+// this is needed until after API1 is retired and pool objects are no longer
+// managed by the GC.
+static void proxy_gc_timerpoke(evutil_socket_t fd, short event, void *arg) {
+    LIBEVENT_THREAD *t = arg;
```

Comment:
Note: 

- set a flag so timerpoke does nothing if the GC did something recently
- check the result of lua_gc in timerpoke and reset proxy_vm_last_kb if it finished a routine